app/questionnaire/navigator.py

```diff
@@ -229,9 +240,69 @@ def get_latest_location(self, completed_blocks=None):
         :return:
         """
         if completed_blocks:
-            incomplete_blocks = [item for item in self.location_path if item not in completed_blocks]
+            incomplete_blocks = [item for item in self.get_location_path() if item not in completed_blocks]
 
             if incomplete_blocks:
                 return incomplete_blocks[0]
 
-        return self.first_location
+        first_location = {
+            "group_id": SchemaHelper.get_first_group_id(self.survey_json),
+            "group_instance": 0,
+            "block_id": SchemaHelper.get_first_block_id(self.survey_json),
+        }
+        return first_location
+
+    def get_front_end_navigation(self, completed_blocks, group_id, group_instance):
+        """
+        Returns the frontend navigation based on the completed blocks, group id and group instance
+
+        :param completed_blocks:
+        :param group_id:
+        :param group_instance:
+        :return: navigation
+        """
+        if 'navigation' not in self.survey_json:
+            return None
+
+        navigation = []
+
+        for group in self.survey_json['groups']:
+
+            logger.debug("Building frontend navigation for group %s", group)
+
+            repeating_rule = SchemaHelper.get_repeating_rule(group)
+
+            if 'hide_in_navigation' not in group:
```

Comment:
Rather than nest this all it might be cleaner to `continue` if it is in the group or change the for loop to exclude these e.g.

```python 
for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
...
```